src/umls.jl

```diff
@@ -0,0 +1,336 @@
+const rest_uri = "https://uts-ws.nlm.nih.gov"
+#----------------------------------------------
+
+#-------------- Common Utils--------------------
+struct BadResponseException <: Exception
+    code::Int64
+end
+
+function Base.showerror(io::IO, e::BadResponseException)
+    print(io, "BadResponseException with code: ",STATUS_CODES[e.code])
+end
+
+#-------------- Authentication -------------------
+"""
+time_to_last_save(file)
+
+Get how many hours since last save 
+"""
+function time_to_last_save(file)
+#unix time is GMT
+time_diff = Dates.value(Dates.unix2datetime(time()) - Dates.unix2datetime(mtime(file)))/ (1000 * 60 * 60)
+return time_diff
+end
+
+"""
+tgt_exists(; tgt_file = "UTS_TGT.txt", hours = 1)
+
+Check if there is a TGT in disk and if it has not expired
+"""
+function tgt_exists(; tgt_file = "UTS_TGT.txt", hours = 1)
+
+if isfile(tgt_file)
+    #check time
+    time_elapsed = time_to_last_save(tgt_file)
+    # Expiration time should be 8 hours - but I tend to expirience bad TGT after few hours
+    if time_elapsed > hours
+        info("UTS TGT Expired")
+        rm(tgt_file)
+        return false
+    else
+        return true
+    end
+end
+
+return false
+end
+
+
+"""
+get_tgt(; force_new::Bool = false, kwargs...)
+
+Retrieve a ticket granting ticket (TGT) using 
+
+1. UTS username and password OR
+2. apikey
+
+A tgt is valid for 8 hours. Therefore, look for UTS_TGT.txt in the local
+directory to see if it has been recently stored. One can force getting a 
+new ticket by passing keyword argument `force_new=true` 
+
+####Examples
+
+```julia
+tgt = get_tgt(username = "myuser", password = "mypass")
+```
+
+```julia
+tgt = get_tgt(apikey = "mykey")
+```
+
+"""
+function get_tgt(; force_new::Bool = false, kwargs...)
+
+    params = Dict(kwargs)
+
+    auth_endpoint = "/cas/v1/tickets/"
+
+    if !haskey(params, :apikey)
+        if !haskey(params, :username) && !haskey(params, :password)
+            error("UMLSRestAPI: Connection requieres passing keword arguments: apikey or username and password")
+        end
+    else
+        auth_endpoint = "/cas/v1/api-key"
+    end
+
+    # Check if there is a valid ticket on disk
+    tgt_file = "UTS_TGT.txt"    
+    if tgt_exists(tgt_file=tgt_file) && !force_new
+        info("UTS: Reading TGT from file")        
+        return readline(tgt_file)
+    end
+
+    info("UTS: Requesting new TGT")
+    headers = Dict("Content-type"=> "application/x-www-form-urlencoded",
+    "Accept"=> "text/plain", "User-Agent"=>"julia" )
+    r = post(uri*auth_endpoint,data=params,headers=headers)
+    ascii_r = String(r.data)
+
+    doc = parsehtml(ascii_r)
+    #for now - harcoded
+    #TO DO:: parse and check
+    try
+        ticket = getattr(doc.root.children[2].children[2], "action")
+
+        open(tgt_file, "w") do f
+            write(f, ticket)
+        end
+
+        return ticket
+    catch
+        error("Could not get TGT: UTS response structure is wrong")
```

Comment:
An exception may happen when opening and writing a ticket to a file in the try block. It should be distinguished from the invalid response.